proxy/http/HttpSM.cc

```diff
@@ -4860,56 +4868,99 @@ HttpSM::do_http_server_open(bool raw)
+      static_cast<TSServerSessionSharingMatchType>(t_state.txn_conf->server_session_sharing_match);
+    if (TS_SERVER_SESSION_SHARING_MATCH_NONE == mt) {
+      mt = static_cast<TSServerSessionSharingMatchType>(t_state.http_config_param->oride.server_session_sharing_match);
+      if (TS_SERVER_SESSION_SHARING_MATCH_NONE == mt) {
+        // Really? Every single session is private *and* outbound connection tracking is on?
+        mt = TS_SERVER_SESSION_SHARING_MATCH_IP; // so use this, least worst choice.
+      }
+    }
+
+    t_state.conn_tracker_info = OutboundConnTracker::get(t_state.current.server->dst_addr, hostname_hash, mt);
+  }
 
-      // if we were previously queued, or the queue is disabled-- just reschedule
-      if (t_state.origin_request_queued || t_state.txn_conf->origin_max_connections_queue < 0) {
+  // Check to see if we have reached the max number of connections on this upstream host.
+  if (t_state.txn_conf->origin_max_connections > 0) {
+    auto &ct_info = *t_state.conn_tracker_info;
+    std::time_t lat;                                   // last alert time.
+    auto ccount                    = ++ct_info._count; // cache instantaneous value plus reservation for this.
+    t_state.outbound_conn_reserved = true;
+    if (ccount > t_state.txn_conf->origin_max_connections) {
+      --ct_info._count; // cancel our reservation
+      t_state.outbound_conn_reserved = false;
+
+      ink_assert(pending_action == nullptr); // in case of reschedule must not have already pending.
+
+      // If the queue is disabled, reschedule.
+      if (t_state.txn_conf->origin_max_connections_queue < 0) {
+        ++ct_info._rescheduled;
         pending_action = eventProcessor.schedule_in(this, HRTIME_MSECONDS(100));
-        return;
-      } else if (t_state.txn_conf->origin_max_connections_queue > 0) { // If we have a queue, lets see if there is a slot
-        ConnectionCountQueue *waiting_connections = ConnectionCountQueue::getInstance();
+      } else if (t_state.txn_conf->origin_max_connections_queue > 0) { // queue enabled, check for a slot
+        auto wcount = ++ct_info._in_queue;
         // if there is space in the queue
-        if (waiting_connections->getCount(t_state.current.server->dst_addr, hostname_hash,
-                                          (TSServerSessionSharingMatchType)t_state.txn_conf->server_session_sharing_match) <
-            t_state.txn_conf->origin_max_connections_queue) {
-          t_state.origin_request_queued = true;
-          Debug("http", "[%" PRId64 "] queued for this host: %s", sm_id,
-                ats_ip_ntop(&t_state.current.server->dst_addr.sa, addrbuf, sizeof(addrbuf)));
-          waiting_connections->incrementCount(t_state.current.server->dst_addr, hostname_hash,
-                                              (TSServerSessionSharingMatchType)t_state.txn_conf->server_session_sharing_match, 1);
+        if (wcount < t_state.txn_conf->origin_max_connections_queue) {
```

Comment:
I think we should remove the queue and return and error response to the client instead of rescheduling.

Reply:
That's a breaking change which I think should be discussed on the dev list. I tracked it down and the rescheduling behavior dates back to before the initial commit to git. Thomas Jackson added the max queue length option to ameliorate the problem.